error_handler.go

```diff
@@ -0,0 +1,198 @@
+// Copyright 2020 The LevelDB-Go and Pebble Authors. All rights reserved. Use
+// of this source code is governed by a BSD-style license that can be found in
+// the LICENSE file.
+
+package pebble
+
+import (
+	"os"
+	"sync"
+
+	"github.com/cockroachdb/errors"
+	"github.com/cockroachdb/pebble/internal/base"
+	"github.com/cockroachdb/pebble/vfs"
+)
+
+// BackgroundErrorReason is an enum of possible operations whose failure may
+// put the DB in read-only mode depending on the severity.
+type BackgroundErrorReason uint8
+
+const (
+	// BgFlush is for errors during background flushes.
+	BgFlush BackgroundErrorReason = iota
+	// BgCompaction is for errors during background compactions.
+	BgCompaction
+
+	// TODO: Export these once we stop panicking for them.
+	// BgWrite is for errors in writing to WAL.
+	bgWrite
+	// BgMemtable is for errors in writing to memtable.
+	bgMemtable
+	// BgManifestWrite is for errors in writing to MANIFEST.
+	bgManifestWrite
+)
+
+// Severity of an error indicates whether recovery is possible and whether DB
+// will be placed in read-only mode or not.
+type Severity uint8
+
+const (
+	// SeverityNoError does not set the background error at all.
+	SeverityNoError Severity = 0
+	// SeveritySoftError does not place the DB in read-only mode and auto recovery
+	// is possible for some of the errors.
+	SeveritySoftError = 1
+	// SeverityHardError places the DB in read-only mode and recovery might be
+	// possible without needing to close then reopen the DB.
+	SeverityHardError = 2
+	// SeverityFatalError places the DB in read-only mode and recovery requires
+	// closing then reopening the DB.
+	SeverityFatalError = 3
+	// SeverityUnrecoverableError places the DB in read-only mode and could mean
+	// data loss, recovery from which is not possible.
+	SeverityUnrecoverableError = 4
+)
+
+// BackgroundError captures the error occurred during any background operation
+// along with its severity. An instance of this is passed in
+// EventListener.BackgroundError.
+type BackgroundError struct {
+	err      error
+	severity Severity
+	reason   BackgroundErrorReason
+}
+
+// Reason returns the operation during which the error occurred.
+func (b *BackgroundError) Reason() BackgroundErrorReason {
+	return b.reason
+}
+
+// Severity returns the severity of the error.
+func (b *BackgroundError) Severity() Severity {
+	return b.severity
+}
+
+// Unwrap returns the error that occurred during the background operation.
+func (b *BackgroundError) Unwrap() error {
+	return b.err
+}
+
+func (b *BackgroundError) Error() string {
+	if b.err != nil {
+		return b.err.Error()
+	}
+	return ""
+}
+
+// Override results in ignoring this background error not placing the DB in
+// read-only mode.
+func (b *BackgroundError) Override() {
```

Comment:
Do we want this override capability? This was introduced in https://github.com/facebook/rocksdb/pull/2477. I don't think the semantics of continuing after an error can be well defined.